File: apps/tauri/src/app/lifecycle/events/projects.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import { toast } from "../../../ui/index";
import { setProjectDialogsStore } from "../../../features/projects/dialogs";
import {
  applyCandidates,
  applyShareActions,
  applySharesApplied,
  applyShareStatus,
  setSharesError,
} from "../../../features/projects/sharesStore";
import type {
  ShareAction,
  ShareCandidate,
  ShareStatusEntry,
  ShareTrigger,
} from "../../../contracts/runtime";
import type { Failure } from "../workbenchAnswer";

export function bindProjectsEvents(): Promise<UnlistenFn[]> {
  return Promise.all([
    // Candidates and rules belong to a project, so
    // they carry no workspace tag to check; status and plans do.
    listen<{ project: string; candidates: ShareCandidate[]; truncated: boolean }>(
      "workbench:share_candidates",
      ({ payload }) => applyCandidates(payload.project, payload.candidates, payload.truncated),
    ),
    listen<Failure>("workbench:share_candidates_failed", ({ payload }) =>
      setSharesError(payload.error),
    ),
    listen<{ workspace: string | null; actions: ShareAction[] }>(
      "workbench:share_plan",
      ({ payload }) => {
        if (payload.workspace) applyShareActions(payload.workspace, payload.actions);
      },
    ),
    listen<Failure>("workbench:share_plan_failed", ({ payload }) => setSharesError(payload.error)),
    listen<{ workspace: string; entries: ShareStatusEntry[] }>(
      "workbench:share_status",
      ({ payload }) => applyShareStatus(payload.workspace, payload.entries),
    ),
    listen<Failure>("workbench:share_status_failed", ({ payload }) =>
      setSharesError(payload.error),
    ),
    listen<Failure>("workbench:share_store_failed", ({ payload }) => setSharesError(payload.error)),
    // Provisioning acks when it *starts*; this is where a worktree stops
    // saying "setting up" and says what it got.
    listen<{
      workspace: string;
      trigger: ShareTrigger;
      actions: ShareAction[];
      error: string | null;
    }>("runtime:shares_applied", (event) => {
      applySharesApplied(
        event.payload.workspace,
        event.payload.trigger,
        event.payload.actions,
        event.payload.error,
      );
      const fell = event.payload.actions.filter((action) => action.fallback).length;
      if (fell > 0) {
        toast({
          title: fell === 1 ? "One file was copied, not cloned" : `${fell} files were copied`,
          detail: "This volume has no copy-on-write, so the clone fell back to a full copy.",
        });
      }
    }),
    listen<{ workspace: string; reason: string }>("runtime:worktree_blocked", (event) => {
      setProjectDialogsStore("worktreeRemoval", {
        workspace: event.payload.workspace,
        reason: event.payload.reason,
      });
    }),
    listen<{ project: string; branch: string; reason: string }>(
      "runtime:worktree_create_failed",
      (event) => setProjectDialogsStore("worktreeCreationFailure", event.payload),
    ),
  ]);
}

File: apps/tauri/src-tauri/src/runtime/workbench/projects.rs
use client::Client;
use domain::{ProjectId, ShareAction, ShareCandidate, ShareRuleId, ShareStatusEntry, WorkspaceId};
use serde::Serialize;
use tauri::AppHandle;

use super::{emit, fail};

pub(super) fn detect_share_candidates(app: &AppHandle, client: &Client, project: ProjectId) {
    match client.detect_share_candidates(project) {
        Ok((candidates, truncated)) => emit(
            app,
            "workbench:share_candidates",
            &CandidatesPayload {
                project,
                candidates,
                truncated,
            },
        ),
        Err(error) => fail(app, "workbench:share_candidates_failed", None, &error),
    }
}

pub(super) fn preview_shares(app: &AppHandle, client: &Client, workspace: WorkspaceId) {
    match client.preview_shares(workspace) {
        Ok(actions) => emit(
            app,
            "workbench:share_plan",
            &PlanPayload {
                workspace: Some(workspace),
                actions,
            },
        ),
        Err(error) => fail(app, "workbench:share_plan_failed", Some(workspace), &error),
    }
}

pub(super) fn load_share_status(app: &AppHandle, client: &Client, workspace: WorkspaceId) {
    match client.share_status(workspace) {
        Ok(entries) => emit(
            app,
            "workbench:share_status",
            &StatusPayload { workspace, entries },
        ),
        Err(error) => fail(
            app,
            "workbench:share_status_failed",
            Some(workspace),
            &error,
        ),
    }
}

pub(super) fn apply_shares(
    app: &AppHandle,
    client: &Client,
    workspace: WorkspaceId,
    only: Option<Vec<ShareRuleId>>,
) {
    // Acks when the work starts; the answer arrives as `SharesApplied`
    // on the event stream, like a fetch.
    if let Err(error) = client.apply_shares(workspace, only) {
        fail(app, "workbench:share_plan_failed", Some(workspace), &error);
    }
}

pub(super) fn adopt_into_share_store(
    app: &AppHandle,
    client: &Client,
    project: ProjectId,
    path: String,
) {
    if let Err(error) = client.adopt_into_share_store(project, path) {
        fail(app, "workbench:share_store_failed", None, &error);
    }
}

pub(super) fn materialize_from_share_store(
    app: &AppHandle,
    client: &Client,
    project: ProjectId,
    path: String,
    workspace: Option<WorkspaceId>,
) {
    if let Err(error) = client.materialize_from_share_store(project, path, workspace) {
        fail(app, "workbench:share_store_failed", workspace, &error);
    }
}

#[derive(Serialize)]
struct CandidatesPayload {
    project: ProjectId,
    candidates: Vec<ShareCandidate>,
    truncated: bool,
}

#[derive(Serialize)]
struct PlanPayload {
    workspace: Option<WorkspaceId>,
    actions: Vec<ShareAction>,
}

#[derive(Serialize)]
struct StatusPayload {
    workspace: WorkspaceId,
    entries: Vec<ShareStatusEntry>,
}
